### PR TITLE
implement `find_in_set`

### PR DESCRIPTION
**Pull request by JCOR11599**
MySQL docs: https://dev.mysql.com/doc/refman/8.0/en/string-functions.html#function_find-in-set

Note: Collations priority isn't exactly the same as MySQL

fix for: https://github.com/dolthub/dolt/issues/5659

### CODE REVIEW

File: sql/expression/function/find_in_set.go [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+		return nil, err
+	}
+	l := strings.ToLower(lVal.(string))
+
+	// always returns 0 when left contains a comma
+	if strings.Contains(l, ",") {
+		return 0, nil
+	}
+
+	// TODO: SET type uses binary arithmetic?
```
Comment (fulghum) [inline sql/expression/function/find_in_set.go:105]: It seems like you're pretty close to supporting set types. If `right`'s type is a SetType, then you could grab right's value and call `right.Type().BitsToString(rVal);` or something, to get a comma separated string representation of the bitmap values mapped back to the set strings. 

This TODO currently isn't going to be very helpful to future code readers though. Try seeing if you can get the simple approach for set types to work here and then this TODO could be more helpful if it were something like "Optimize set type lookup to work off of bitmap directly" or something. That would give future readers a better clue about what needs to be done in the future.

File: sql/expression/function/find_in_set_test.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			expected: 3,
+		},
+		{
+			name:     "special characters ok",
+			left:     "test@example.com",
+			right:    "nottest@example.com,hello@example.com,TEST@eXaMpLe.COM",
+			expected: 3,
+		},
+		{
+			name:     "look for empty string",
```
Comment (fulghum) [inline sql/expression/function/find_in_set_test.go:84]: Good edge case! 🙌

File: sql/expression/function/find_in_set.go [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+
+	if left == nil || right == nil {
+		return nil, nil
+	}
+
+	lVal, _, err := types.LongText.Convert(left)
+	if err != nil {
+		return nil, err
+	}
+	l := strings.ToLower(lVal.(string))
```
Comment (fulghum) [inline sql/expression/function/find_in_set.go:98]: `find_in_set` is not actually case insensitive, so we shouldn't be lower casing here... 

Whether the match is case sensitive or not comes down to the collation in use. Here's an example with MySQL:

```
mysql> show variables like 'collation_%';
+----------------------+--------------------+
| Variable_name        | Value              |
+----------------------+--------------------+
| collation_connection | utf8mb4_0900_ai_ci |
| collation_database   | utf8mb4_0900_ai_ci |
| collation_server     | utf8mb4_0900_ai_ci |
+----------------------+--------------------+
3 rows in set (0.00 sec)

mysql> select find_in_set('A', 'a,b,c');
+---------------------------+
| find_in_set('A', 'a,b,c') |
+---------------------------+
|                         1 |
+---------------------------+
1 row in set (0.00 sec)

mysql> set @@collation_connection='utf8mb4_0900_bin';
Query OK, 0 rows affected (0.01 sec)

mysql> select find_in_set('A', 'a,b,c');
+---------------------------+
| find_in_set('A', 'a,b,c') |
+---------------------------+
|                         0 |
+---------------------------+
1 row in set (0.00 sec)
```

**Review by fulghum — COMMENTED** [review]
Looks very close. I think you can easily add in support for set types (but skip the more optimized direct bitmap comparing improvements for now) and we need to use the collation for matching in order to get the right case sensitive/insensitive behavior.

File: enginetest/queries/queries.go [hunk 4]
```diff
@@ -7500,6 +7500,28 @@ SELECT * FROM my_cte;`,
 			{4},
 		},
 	},
+	{
```
Comment (fulghum) [inline enginetest/queries/queries.go:7503]: It looks like all the tests we run are testing against a string type. You should add a test that calls `find_in_set` on an enum column and on a set column so that we have coverage over those cases, too.

File: sql/expression/function/find_in_set.go [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+	}
+
+	collationPreference, _ := f.CollationCoercibility(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	strType := types.CreateLongText(collationPreference)
+	for i, r := range strings.Split(r, ",") {
+		cmp, err := strType.Compare(l, r)
```
Comment (fulghum) [inline sql/expression/function/find_in_set.go:129]: @Hydrocharged – is there something else we need to do to get collation aware comparisons? It seems like the string type would encapsulate that if we have the right collation preference set here, but James is still seeing the case insensitive collation test cases not match like we are expecting them to.

**Review by fulghum — COMMENTED** [review]
Thanks for adding in the support for sets. Make sure there's a test covering that code path – I didn't see one, but might have missed it. Would be good to add a test case for enums, too. I think the current code will handle them correctly because of the LongText conversion, but would be good to throw in a quick test when you add the set test. 

I also tagged @Hydrocharged to see if he has advice on the collation issue in the skipped test. It seems like we should be able to make that work, or create a tracking issue if there's a bug fix for collation logic needed that's separate from this PR.

File: sql/expression/function/find_in_set.go (hunk 5 above)
Comment (Hydrocharged) [reply sql/expression/function/find_in_set.go:133]: Hey Jason, I talked to James about this. His implementation of `CollationCoercibility` doesn't take into account the left and right expressions, so I gave him some pointers on how to resolve it.

**Review by fulghum — APPROVED** [review]
Nice work! 🚢
